### ISSUE TITLE
Batch remove participants from Sakai sites

### ISSUE THREAD

**Issue by joliet** [DukeLearningInnovation/kits#529]
I am academic technology support staff and trying to bulk remove participants from Sakai sites but that feature doesn't exist, which makes me feel overwhelmed with the manual process of removing people from many sites.

How might we (the Kits project team) make it possible for academic technology support staff to bulk remove users from Sakai sites?

**Questions**

**Assumptions**
The primary use case for this is removing staff members from Sakai sites.

**Hypotheses**
We will add a new feature to the Kits Admin Console that batch removes users from Sakai sites.

**Comment by joliet** [comment]
December 1 would be a great deadline for this feature.  UI could likely use existing components.

**Comment by joliet** [comment]
@ewendland23 and @noahhh .  Jen and I discussed this one today.  The workflow is exactly the same as Batch Add, just with the remove action instead.  Details are here - https://docs.google.com/document/d/18uQHBn0p_TG8ittKnPZRtXwQNNz2nvORB66CPkkzfg0/edit?usp=sharing

We could either make it one feature with two options (select add or remove), or we could make them two separate options in the Admin console with their own icons.  Do you have a preference from a technical perspective?

**Comment by joliet** [comment]
Reach out to jlbecker@duke.edu (Jennifer Becker) and Karin Reese about the design for this one.

**Comment by joliet** [comment]
Here's a non-pixel perfect mockup of a design for this.  Need feedback from @noahhh and @curiousmustard 

![Animated GIF-source (2)](https://user-images.githubusercontent.com/7387760/127545742-fd2bfaf3-4ad7-43d6-a46d-120ba23d87a9.gif)

**Comment by noahhh** [comment]
@joliet how does this work for combined kits?  Do we include it on the edit page for them?  With a single kit we know which course we are removing them from

**Comment by noahhh** [comment]
@joliet I had time to take a look at this and make some changes for it, it is out in test.  Not sure how to handle batch remove on combined kits though (maybe we don't let them?).  I went in to take a look at it and it ended up not being a ton of work to do

**Comment by joliet** [comment]
Hey @noahhh .  Thanks for taking a look at this and working on it.  Apologies about the oversight with combined kits.  I should have provided a mockup for that, too.  Here's what it looks like without the Action selection drop-down:

<img width="698" alt="Batch_Edit_Combined_Kits" src="https://user-images.githubusercontent.com/7387760/127915426-3785cbce-8a2e-43af-90ad-f39573e955c5.png">

We can follow up on this when you're back from vacation.  Have a great trip!

**Comment by joliet** [comment]
Testing update @noahhh.

Tested with Combined Kit COMBINED KIT ARABIC.101.001.F20, kit id 2023

I used a list of participants from the Advising kit on production.  Some of these are current Duke users and some are not.

I used the email address for these users.

When adding to the kit, selecting the first roster (ARABIC 101 - 001: Elementary Arabic (Lecture) ) to add the users to, I do not receive a confirmation message or any feedback on status.  I only see this if I use the accordion to open the ARABIC 101 - 001: Elementary Arabic (Lecture) group shown below the batch add/remove text entry box at the top.

After waiting for about a minute, I get the error 'Page Unresponsive'.

<img width="451" alt="Page_Unresponsive_and_Duke_Kits" src="https://user-images.githubusercontent.com/7387760/133094679-6a53ce8d-154a-456e-b736-f63cf563ed43.png">

Coming back to the participant list after dismissing the error I see that the members I added through batch are in the ARABIC 101 - 001: Elementary Arabic (Lecture) group/roster in the combined kit and see a message to use a Duke email for a subset of members.  Is this case sensitive?  The original email addresses use caps which is what displays when searching for this user in the directory.  The Duke email addresses here are active emails.

<img width="952" alt="Use_Duke_email_batch_add_message" src="https://user-images.githubusercontent.com/7387760/133095615-3a110c08-6cec-4a93-b3fe-7ee835b8dc41.png">

Batch remove produced the following message:

<img width="1364" alt="Use_Duke_email_batch_remove_message" src="https://user-images.githubusercontent.com/7387760/133096012-34a7445f-b337-4374-8e3d-5e404da8a58f.png">

Some of these users were added to the kit despite the error (e.g. Linda.Daniel@duke.edu and Dean.Sue@duke.edu) but were not removed in batch remove.

Now that we have two processes here, add and remove, the error message could be edited to use this language "Please add or remove Duke participants with their email. email@duke.edu is not a recognized email."

I seem to remember Liz stating it could be challenging to use email for adding users since some users have multiple aliases.  Is this still true?  Should we insist on just first name and last name or netid?

Testing with the same user list in the combined kit COMBINED KIT ECON.101D.001.Sp20, kit id 1838:

Testing batch add: users were added to the kit in the ECON 101D - 001: Economic Principles (Lecture) roster/kit with their email addresses.

Testing batch remove: 

Users were not removed using NetID.  Nothing happened.  No error, no visual evidence of activity, and users were still present in the first.

Users were removed using DUID but no feedback on completion, even with the kit open where the members were added - ECON 101D - 001: Economic Principles (Lecture).

**Comment by noahhh** [comment]
@joliet Thanks for testing.  I found an issue, I assume primarily with emails (as you experienced), in which we were downcasing the user query but not the email returned from LDAP.  This resulted in them being found and added, but to generate the error message we see if the LDAP result was in the user query.  Since LDAP was capitalized and the user query was not, it displayed the error message (such as you saw for Linda and Dean).  I'm putting a fix for this in to test now.

How many users were you testing with adding/removing when you saw the page unresponsive?  I wonder if that was a timeout due to the batch process still processing or something.  I have added 5-10 with no problems, but haven't gone beyond that.  Could you send me the list you were using?

I believe it is still true that users can have active/unactive/aliases for emails. I don't know if LDAP only returns a user if we search by their preferred, or any alias.

**Comment by joliet** [comment]
Thanks for the fix @noahhh .  Will test that out shortly.

The list I was using is here - https://docs.google.com/spreadsheets/d/1Y_0bg9d_H7a5dGMcRPud2nMRBABR6CNfbYlE14AEFLI/edit?usp=sharing 

and has 444 members.  You'll need to use your Duke Google account for access.  Let me know if you have any trouble.

**Comment by joliet** [comment]
Are you still working on this @noahhh ?  I just re-tested and got the "Page Unresponsive" error again when bulk adding members.  It was to this kit - LAW 2020-10: Demo for Law (id 2009) and with this member list - https://docs.google.com/spreadsheets/d/1-u5kOd5reQMjFtdaMqoozGxKYaq25jzEVoQRmkDfq4g/edit?usp=sharing

The good news is that the adding and removing process is working!

**Comment by noahhh** [comment]
@joliet I have not made any changes aside from my earlier message.  I think the page unresponsive thing is something I'd like to talk about in our next Kits bi-weekly meeting if that is alright.

**Comment by joliet** [comment]
Sure, no problem @noahhh !  Let's chat next week about this one.

**Comment by noahhh** [comment]
@joliet I changed it so batch add/remove members kicks off a background process and sends an email upon completion.  This is out in test.

**Comment by noahhh** [comment]
@joliet just saw an error come through, I was calling our logger incorrectly.  Pushing a small change to test

**Comment by joliet** [comment]
@noahhh Yeah, I made a mistake.  I selected NetID in the radio buttons but then entered email addresses.  This is good because mistakes will happen.

I have one more request.  After the snack bar confirmation message about results going to email, could we clear the text box?  Seeing what I entered there makes me feel like the task is unfinished even though I received the confirmation.  

I think we have another usability issue here.  Once the task is over there's no clear way to go back to the original view.  If you click Edit Mode again, you lose the page header (Roster) and top-level navigation link (BACK TO MY KITS).  It feels weird to me, but I should confirm through a usability test what users would expect to do *after* bulk edit.  Maybe staying in this view is fine with our users or maybe they would just go back to the My Kits page since the task is done.  I'll be asking a few folks to test this out before we move to prod and this is a question I will ask.

**Comment by noahhh** [comment]
@joliet I put out a change that will clear the input box when you submit.

What do you mean the 'original view'?  If you click 'Cancel' it brings back the original, single user search, if that is what you mean.
